### PR TITLE
[DRAFT] Support MMIO regions

### LINKED ISSUE #1304: How to verify mmio access?
I've tried with the below code for mocking a mmio access.

```
const BUFFER: *mut u32 = 0xb8000 as *mut u32;
#[cfg(kani)]
#[kani::proof]
fn test_write() {
    let val = 12;
    unsafe {
        core::ptr::write_volatile(BUFFER, val);
    }
}
```

using the following command line invocation:

```
cargo kani
```

with Kani version: 0.3.0

It bought out the following failures. 
```
SUMMARY:
 ** 5 of 7 failed
Failed Checks: dereference failure: pointer NULL
 File: "/rustc/c52b9c10bfb5164015eb977ff498e0597ae63eb1/library/core/src/ptr/mod.rs", line 1519, in std::ptr::write_volatile
Failed Checks: dereference failure: deallocated dynamic object
 File: "/rustc/c52b9c10bfb5164015eb977ff498e0597ae63eb1/library/core/src/ptr/mod.rs", line 1519, in std::ptr::write_volatile
Failed Checks: dereference failure: dead object
 File: "/rustc/c52b9c10bfb5164015eb977ff498e0597ae63eb1/library/core/src/ptr/mod.rs", line 1519, in std::ptr::write_volatile
Failed Checks: dereference failure: pointer outside object bounds
 File: "/rustc/c52b9c10bfb5164015eb977ff498e0597ae63eb1/library/core/src/ptr/mod.rs", line 1519, in std::ptr::write_volatile
Failed Checks: dereference failure: invalid integer address
 File: "/rustc/c52b9c10bfb5164015eb977ff498e0597ae63eb1/library/core/src/ptr/mod.rs", line 1519, in std::ptr::write_volatile

VERIFICATION:- FAILED
```

Why did the verification failures occur and how to pass the verification when an object of absolute address (e.g., mmio) is involved?

### PR DESCRIPTION
**Pull request by danielsn**
### Description of changes: 

Adds support for MMIO regions to kani proof harnesses.

### Resolved issues:

Resolves #1304 

### Call-outs:

- [ ] This is dependent on https://github.com/diffblue/cbmc/pull/6747. It will not work until that is released. I tested against a locally built copy.
- [ ] We need documentation for this feature
- [x] Should probably add some negative tests too
- [ ] Should test that it behaves like MMIO: writes don't need to be the same as is read back, two reads can get different values

### Testing:

* How is this change tested? New regression tests.

* Is this a refactor change? No

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by kroening** [comment]
Why not simply have `--mmio_region` enabled by default?

**Comment by kroening** [comment]
> Why not simply have `--mmio_region` enabled by default?

(Ignore, it is already)

File: tests/kani/MMIO/semantics.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    let val2 = 314;
+    unsafe {
+        let p1 = BUFFER;
+        let p2 = BUFFER.offset(2);
+        *p1 = val1;
+        *p2 = val2;
+        assert_eq!(*p1, val1);
+        assert_eq!(*p2, val2);
+        assert_eq!(*p2, val1);
+        assert_eq!(*p1, val2);
```
Comment (zpzigi754) [inline tests/kani/MMIO/semantics.rs:61]: I guess that these should be changed like 
```
assert_ne(*p2, val1); 
assert_ne(*p1, val2);
```

File: library/kani_macros/src/lib.rs [hunk 2]
```diff
@@ -73,3 +73,26 @@ pub fn unwind(attr: TokenStream, item: TokenStream) -> TokenStream {
...
+#[cfg(not(kani))]
+#[proc_macro_attribute]
+pub fn mmio_region(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    // When the config is not kani, we should leave the function alone
+    item
+}
+
+/// Set Loop unwind limit for proof harnesses
+/// The attribute '#[kani::unwind(arg)]' can only be called alongside '#[kani::proof]'.
+/// arg - Takes in a integer value (u32) that represents the unwind value for the harness.
```
Comment (fzaiser) [inline library/kani_macros/src/lib.rs:86]: Update comment

File: kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs [hunk 3]
```diff
@@ -297,16 +297,39 @@ impl<'tcx> GotocCtx<'tcx> {
...
+                        return;
+                    }
+                }
+            }
+        }
+
+        // If we didn't successfully process, give an error message
+        self.tcx
+            .sess
+            .span_err(attr.span, "Expected exactly two integers are the argument to 'mmio_region'");
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:322]: Can you print or at least log which values we got?

File: kani_metadata/src/harness.rs [hunk 4]
```diff
@@ -16,4 +16,6 @@ pub struct HarnessMetadata {
     pub original_line: String,
     /// Optional data to store unwind value
     pub unwind_value: Option<u32>,
+    /// Optional MMIO regions
+    pub mmio_regions: Vec<(u128, u128)>,
```
Comment (celinval) [inline kani_metadata/src/harness.rs:20]: Should we use `usize` instead?

File: tests/expected/mmio/expected [hunk 5]
```diff
@@ -0,0 +1,26 @@
+Status: SUCCESS\
+Description: "offset: attempt to compute number in bytes which would overflow"
```
Comment (celinval) [inline tests/expected/mmio/expected:2]: This seems irrelevant for your test. I would just remove it.

File: tests/expected/mmio/main.rs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+#[cfg(kani)]
+#[kani::proof]
+#[kani::mmio_region(0x8000, 8)]
+fn test_write() {
+    let val = 12;
+    unsafe {
+        //BUFFER+2 is not in the MMIO region. Expect pointer check failures.
+        *(BUFFER.offset(2)) = val;
+        //BUFFER2 is not in the MMIO region. Expect pointer check failures.
+        *BUFFER2 = val;
```
Comment (celinval) [inline tests/expected/mmio/main.rs:16]: You are not really checking that both fail though. This test will succeed if only one of these fail. You can either add the line in the file to the expected file or break this into two harness and add the function name. I prefer the second since it is always painful to maintain LoC up to date.

File: tests/expected/mmio/main.rs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+const BUFFER: *mut u32 = 0x8000 as *mut u32;
+const BUFFER2: *mut u32 = 0x1000 as *mut u32;
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::mmio_region(0x8000, 8)]
+fn test_write() {
```
Comment (celinval) [inline tests/expected/mmio/main.rs:10]: can you rename this to `check_invalid_write()` or something like that?

File: tests/kani/MMIO/semantics.rs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+#[kani::mmio_region(0x8000, 4)]
+fn test_read_is_stable() {
+    unsafe {
+        let val1 = core::ptr::read_volatile(BUFFER);
+        let val2 = core::ptr::read_volatile(BUFFER);
+        assert_eq!(val1, val2);
+    }
+}
+
+//TODO Weirdly, these fail even tho the other tests pass???
```
Comment (celinval) [inline tests/kani/MMIO/semantics.rs:29]: This looks like a blocker to me. Maybe we should investigate this behavior a bit better.

File: tests/kani/MMIO/semantics.rs [hunk 9]
```diff
@@ -0,0 +1,63 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+const BUFFER: *mut u32 = 0x8000 as *mut u32;
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::mmio_region(0x8000, 4)]
+fn test_read_after_write_doesnt_havoc() {
+    let val = 12;
```
Comment (celinval) [inline tests/kani/MMIO/semantics.rs:10]: Can you use `any` instead?

File: tests/kani/MMIO/semantics.rs [hunk 10]
```diff
@@ -0,0 +1,63 @@
+// Copyright Kani Contributors
```
Comment (celinval) [inline tests/kani/MMIO/semantics.rs:1]: Can you please document what these tests are testing?

File: tests/kani/MMIO/test.rs [hunk 11]
```diff
@@ -0,0 +1,39 @@
+// Copyright Kani Contributors
```
Comment (celinval) [inline tests/kani/MMIO/test.rs:1]: Can you cover 0 sized operations?

File: tests/kani/MMIO/semantics.rs [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        core::ptr::write_volatile(BUFFER, val);
+        let new_val = core::ptr::read_volatile(BUFFER);
+        assert_eq!(new_val, val);
+    }
+}
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::mmio_region(0x8000, 4)]
+fn test_read_is_stable() {
```
Comment (celinval) [inline tests/kani/MMIO/semantics.rs:21]: But should it be stable?

**Review by celinval — CHANGES_REQUESTED** [review]
The semantics here aren't clear to me. Your `test_read_is_stable` contradicts the item in the callout:

> Should test that it behaves like MMIO: writes don't need to be the same as is read back, two reads can get different values.

**Comment by celinval** [comment]
@danielsn any news here?
